### DISCUSSION TITLE
Electron + Scanner integration using proprietary DLL

### DISCUSSION THREAD

**Discussion by jtommy** [agracio/edge-js#195] [Q&A] [answered]
Hi!
I'm integrating in our desktop application (electron 10.x) a proprietary dll that allows us to manage a scanner usb. 
The dll is in C# and it looks as wrapper of C dll (indeed i have to put an other dll on system32 or in the same folder of the applicaiton).

As usual I mapped all functions that i need from dll as edge.function as below:

My code:

```js
import edge from 'electron-edge-js';
import path from 'path';

declare const __static: string;

const dllA6 = path.join(process.env.NODE_ENV === 'production' ? __dirname : __static, '../src/dll/CuDoubleSidedScannerAPIWrap.dll');


export class ScannerA6 {
      Init = () => edge.func<string, string[]>(`
	  	#r "${dllA6}"
		using Custom.CuDoubleSidedScannerAPIWrap;
		using Custom.CuDoubleSidedScannerAPIWrap.ScannerEvent;
		async (input) => {
			CuDoubleSidedScannerAPI cScannerLibrary = new CuDoubleSidedScannerAPI();
			return cScannerLibrary.EnumScanner();
		}
	`);

      Status = () => edge.func<string, object>(`
		#r "${dllA6}"
		using Custom.CuDoubleSidedScannerAPIWrap;
		using Custom.CuDoubleSidedScannerAPIWrap.ScannerEvent;
		async (input) => {
			CuDoubleSidedScannerAPI cScannerLibrary = new CuDoubleSidedScannerAPI();
			CuScanner scanner = cScannerLibrary.ConnectScanner(input.ToString());
			return scanner.GetStatus();
		}
	`);

}
```

The first time that i run the application i can call Init but when i try to call Status or Init again the c# code throws the following exception:

From scanner docs
```
ERR_LIBRARY_ALREADY_INITIALIZED 	
The library has already been initialized. Please deinitialize the library before continue
```

I tried to decompile it but the code it looks like obfuscated and i cannot know the real reason. 

Is there a way to reset completely the edge env or a similar thing in order to avoid that exception and every call is like the first?

Thanks!

PS: thanks @agracio because i already use your project for a similar integration it works very well!

**Comment by jtommy** [comment]
@agracio maybe using `import('edge-js')` or `require('edge-js)` into each methods (Init, Status, ecc)?

**Comment by agracio** [comment]
This looks like an error coming from your application and not from Edge.js, have you tried using this code in c# console app to see if it works in this way? `require('edge-js)` has nothing to do with this error from the looks of it.

**Reply by jtommy** [reply]
What do you mean? The dll is not my code, it comes from scanner provider.
I have a complete example of the scanner provider but it's a C# project with GUI and I have my UI and I need only a part of their functions l.

**Reply by jtommy** [reply]
The error  always happens when I create a new instance of `CuDoubleSidedScannerAPI cScannerLibrary = new CuDoubleSidedScannerAPI()` after a first time but I cannot share it because it's a complex object, no?

**Comment by agracio** [comment]
The best advise I can give you is to clone https://github.com/agracio/edge-js-quick-start and then add your code as C# methods rather than inline. This should either resolve your problem or point you to the right place in C# code where the error occurs. Inline code does not offer much insight into what is happening in C#.

**Reply by jtommy** [reply]
Can you make an example relative to my code?

**Reply by agracio** [reply, marked as answer]
In pure C# class you can create `new CuDoubleSidedScannerAPI()` once and then use it in all of your methods, so as I said clone https://github.com/agracio/edge-js-quick-start and create a proper C# class where you instantiate `CuDoubleSidedScannerAPI` once and all the methods use that instance. Quick start provides great example on how to use compiled C# dlls rather than inline C#.

Something like this
```cs
    public class DoubleSidedScanner
    {
        CuDoubleSidedScannerAPI scannerLibrary = new CuDoubleSidedScannerAPI();
        
        public async Task<object> Init(dynamic input)
        {
            return scannerLibrary.EnumScanner();
        }
        
        public async Task<object> Status(dynamic input)
        {
            CuScanner scanner = scannerLibrary.ConnectScanner(input.ToString());
            return scanner.GetStatus();
        }
    }
```

**Reply by agracio** [reply]
And of course there is also https://github.com/agracio/electron-edge-js-quick-start so you can try it in Electron app.

**Reply by jtommy** [reply]
Can i import an external dll in C# class?

**Reply by agracio** [reply]
Yes it is a very basic C#/.NET feature
